Metro server crashes

# Problem

The Metro server is crashing after 1-2 seconds after the server is started.

```log
➜  status-mobile ✗ make run-metro  
Configuring Nix shell for target 'clojure'...
No changes detected.
Done in: 1.159s
yarn run v1.22.19
$ react-native start
warn Package react-native-blob-util contains invalid configuration: "dependency.hooks" is not allowed. Please verify it's properly linked using "react-native config" command and contact the package maintainers about this.
warn Package react-native-navigation contains invalid configuration: "dependency.assets" is not allowed,"dependency.hooks" is not allowed. Please verify it's properly linked using "react-native config" command and contact the package maintainers about this.
                                                      
                        #######                       
                   ################                   
                #########     #########               
            #########             ##########          
        #########        ######        #########      
       ##########################################     
      #####      #####################       #####    
      #####          ##############          #####    
      #####    ###       ######       ###    #####    
      #####    #######            #######    #####    
      #####    ###########    ###########    #####    
      #####    ##########################    #####    
      #####    ##########################    #####    
      #####      ######################     ######    
       ######        #############        #######     
         #########        ####       #########        
              #########          #########            
                  ######### #########                 
                       #########                      
                                                      
                                                      
                    Welcome to Metro!
              Fast - Scalable - Integrated



To reload the app press "r"
To open developer menu press "d"

jest-haste-map: Watchman crawl failed. Retrying once with node crawler.
  Usually this happens when watchman isn't running. Create an empty `.watchmanconfig` file in your project's root folder or initialize a git or hg repository in your project.
  Error: Watchman error: A non-recoverable condition has triggered.  Watchman needs your help!
The triggering condition was at timestamp=1687286390: opendir(/Users/javid/Projects/status-mobile/node_modules/metro-core/node_modules/jest-regex-util/build) -> Too many open files
All requests will continue to fail with this message until you resolve
the underlying problem.  You will find more information on fixing this at
https://facebook.github.io/watchman/docs/troubleshooting.html#poison-opendir. Make sure watchman is running for this project. See https://facebook.github.io/watchman/docs/troubleshooting.
node:events:491
      throw er; // Unhandled 'error' event
      ^

Error: A non-recoverable condition has triggered.  Watchman needs your help!
The triggering condition was at timestamp=1687286390: opendir(/Users/javid/Projects/status-mobile/node_modules/metro-core/node_modules/jest-regex-util/build) -> Too many open files
All requests will continue to fail with this message until you resolve
the underlying problem.  You will find more information on fixing this at
https://facebook.github.io/watchman/docs/troubleshooting.html#poison-opendir

    at ChildProcess.<anonymous> (/Users/javid/Projects/status-mobile/node_modules/fb-watchman/index.js:207:21)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
Emitted 'error' event on WatchmanWatcher instance at:
    at Client.<anonymous> (/Users/javid/Projects/status-mobile/node_modules/metro/node_modules/jest-haste-map/build/watchers/WatchmanWatcher.js:150:10)
    at Client.emit (node:events:513:28)
    at ChildProcess.<anonymous> (/Users/javid/Projects/status-mobile/node_modules/fb-watchman/index.js:214:12)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5) {
  watchmanResponse: {
    version: '2023.01.30.00',
    error: 'A non-recoverable condition has triggered.  Watchman needs your help!\n' +
      'The triggering condition was at timestamp=1687286390: opendir(/Users/javid/Projects/status-mobile/node_modules/metro-core/node_modules/jest-regex-util/build) -> Too many open files\n' +
      'All requests will continue to fail with this message until you resolve\n' +
      'the underlying problem.  You will find more information on fixing this at\n' +
      'https://facebook.github.io/watchman/docs/troubleshooting.html#poison-opendir\n'
  }
}

Node.js v18.16.0
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
make: *** [run-metro] Error 1
```


#### Expected behaviour

The Metro server should work as expected without any errors.


### Additional Information

- Operating System: macOS 13.4
- Xcode version: `v14.3.1`
- Node: `v18.16.0`

Adding `node_modules` to `ignore_dirs` didn't work for me:

<details>

  <summary>make run-metro output</summary>
  
```
To reload the app press "r"
To open developer menu press "d"

jest-haste-map: Watchman crawl failed. Retrying once with node crawler.
  Usually this happens when watchman isn't running. Create an empty `.watchmanconfig` file in your project's root folder or initialize a git or hg repository in your project.
  Error: Watchman error: A non-recoverable condition has triggered.  Watchman needs your help!
The triggering condition was at timestamp=1687360117: opendir(/Users/vkjr/work/projects/status/tmp/status-mobile/node_modules/get-value) -> Too many open files in system
All requests will continue to fail with this message until you resolve
the underlying problem.  You will find more information on fixing this at
https://facebook.github.io/watchman/docs/troubleshooting.html#poison-opendir. Make sure watchman is running for this project. See https://facebook.github.io/watchman/docs/troubleshooting.
Failed to construct transformer:  TypeError: child.send is not a function
    at ChildProcessWorker.initialize (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/workers/ChildProcessWorker.js:173:11)
    at new ChildProcessWorker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/workers/ChildProcessWorker.js:124:10)
    at WorkerPool.createWorker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/WorkerPool.js:44:12)
    at new BaseWorkerPool (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/base/BaseWorkerPool.js:127:27)
    at new WorkerPool (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/WorkerPool.js:30:1)
    at new Worker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/index.js:167:26)
    at HasteMap._getWorker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-haste-map/build/index.js:862:24)
    at HasteMap._processFile (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-haste-map/build/index.js:756:17)
    at HasteMap._buildHasteMap (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-haste-map/build/index.js:809:28)
    at /Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-haste-map/build/index.js:459:33
node:events:491
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at target._send (node:internal/child_process:865:20)
    at target.send (node:internal/child_process:738:19)
    at ChildProcessWorker.initialize (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/workers/ChildProcessWorker.js:173:11)
    at new ChildProcessWorker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/workers/ChildProcessWorker.js:124:10)
    at WorkerPool.createWorker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/WorkerPool.js:44:12)
    at new BaseWorkerPool (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/base/BaseWorkerPool.js:127:27)
    at new WorkerPool (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/WorkerPool.js:30:1)
    at new Worker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-worker/build/index.js:167:26)
    at HasteMap._getWorker (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-haste-map/build/index.js:862:24)
    at HasteMap._processFile (/Users/vkjr/work/projects/status/status-mobile/node_modules/metro/node_modules/jest-haste-map/build/index.js:756:17)
Emitted 'error' event on ChildProcess instance at:
    at node:internal/child_process:869:39
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11) {
  errno: -32,
  code: 'EPIPE',
  syscall: 'write'
}

Node.js v18.16.0
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
make: *** [run-metro] Error 1
```

</details>

--- a/nix/overlay.nix
+++ b/nix/overlay.nix
@@ -49,6 +49,12 @@ in {
     doCheck = false;
   });
 
+  # Downgrade watchman in attempt to fix "too many files open issue"
+  watchman = callPackage ./pkgs/watchman {
+    inherit (super.darwin.apple_sdk.frameworks) CoreServices;
+    autoconf = super.buildPackages.autoconf269;
+  };
+
   # Package version adjustments
   gradle = super.gradle_7;
   nodejs = super.nodejs-18_x;

--- a/nix/pkgs/watchman/README.md
+++ b/nix/pkgs/watchman/README.md
@@ -1,0 +1,23 @@
+# Issue
+
+`watchman` was upgraded significantly during the last nixpkgs upgrade (`4.9.0` (Aug 16, 2017) to `2023.01.30.00` - 6 years between):
+- https://github.com/status-im/status-mobile/pull/14944
+- https://github.com/status-im/nixpkgs/commit/4e9c02bcc709fe1737a746add0e8e0109133d808
+
+Probably causing developers to have "too many files open" issue:
+https://github.com/status-im/status-mobile/issues/16341
+
+```
+Error: A non-recoverable condition has triggered.  Watchman needs your help!
+The triggering condition was at timestamp=1687286390: opendir(/Users/javid/Projects/status-mobile/node_modules/metro-core/node_modules/jest-regex-util/build) -> Too many open files
+All requests will continue to fail with this message until you resolve
+the underlying problem.  You will find more information on fixing this at
+https://facebook.github.io/watchman/docs/troubleshooting.html#poison-opendir
+```
+
+# Fix
+This is an attempt to fix the issue by downgrading the watchman
+
+# Upgrade
+When upgrading in the future, please read the comments:
+https://github.com/status-im/status-mobile/issues/16341

--- a/nix/pkgs/watchman/default.nix
+++ b/nix/pkgs/watchman/default.nix
@@ -1,0 +1,48 @@
+{ stdenv, lib, config, fetchFromGitHub, autoconf, automake, pcre
+, libtool, pkg-config, openssl
+, confFile ? config.watchman.confFile or null
+, withApple ? stdenv.isDarwin, CoreServices
+}:
+
+stdenv.mkDerivation rec {
+  pname = "watchman";
+  version = "4.9.0";
+
+  src = fetchFromGitHub {
+    owner = "facebook";
+    repo = "watchman";
+    rev = "v${version}";
+    sha256 = "0fdaj5pmicm6j17d5q7px800m5rmam1a400x3hv1iiifnmhgnkal";
+  };
+
+  nativeBuildInputs = [ autoconf automake pkg-config libtool ];
+  buildInputs = [ pcre openssl ]
+    #++ lib.optionals withApple [ CoreServices ];
+    ++ lib.optionals stdenv.isDarwin [ CoreServices ];
+
+  configureFlags = [
+    "--enable-lenient"
+    "--enable-conffile=${if confFile == null then "no" else confFile}"
+    "--with-pcre=yes"
+
+    # For security considerations re: --disable-statedir, see:
+    # https://github.com/facebook/watchman/issues/178
+    "--disable-statedir"
+  ];
+
+  prePatch = ''
+    patchShebangs .
+  '';
+
+  preConfigure = ''
+    ./autogen.sh
+  '';
+
+  meta = with lib; {
+    description = "Watches files and takes action when they change";
+    homepage    = "https://facebook.github.io/watchman";
+    maintainers = with maintainers; [ cstrahan ];
+    platforms   = with platforms; linux ++ darwin;
+    license     = licenses.asl20;
+  };
+}

